### ISSUE TITLE
MPI gap_fit

### ISSUE THREAD

**Issue by janklinux** [libAtoms/QUIP#397]
Moi,

I have heard that there is an MPI version of gap_fit in development / available in the current branch. I have tried running an MPI version of gap_fit (from a fresh clone using the provided arch file) but it does not seem to give the expected result in terms of memory parallelism that I expected. I was under the impression, that using the MPI version I could now use two nodes and with that have more (almost double) memory available to fit my potential. But at the moment this does not work for me, I have to reduce training data to about the same amount as I can fit on a single node only so there is no benefit from it. Otherwise the jobs simply crash with memory allocation errors.
Did I understand something wrong? If not, how do I use the MPI version properly?

Thanks, Jan

**Comment by jameskermode** [comment]
This functionality is very new (just merged into `public` this week) and there are some limitations - principally that you need to do a standard run first to generate the sparse points - so for most users we suggest waiting until we can provide some examples and documentation.

However, if you are very keen to try it here are some very rough instructions (adapted from those provided by @Sideboard for local users to test here at Warwick). If you give it a shot please give us feedback on your experiences.

### Compiling

1. Load appropriate modules to provide Fortran, BLAS/LAPACK and ScaLAPACK. On our HPC cluster this requires `module load GCC/10.3.0 OpenMPI/4.1.1 ScaLAPACK/2.1.0-fb OpenBLAS/0.3.15`
2. `QUIP_ARCH=linux_x86_64_gfortran_openmpi+openmp`
3.  `make config` creates `build/$QUIP_ARCH/Makefile.inc`
    * `MATH_LINKOPTS=-lscalapack -lopenblas`
    * `HAVE_GAP=1`
    * `HAVE_QR=1`
    * `HAVE_SCALAPACK=1`

### Using

1. Calculate without MPI to get the sparseX output files
    * Use `sparsify_only_no_fit=T` to just create the sparseX files
3. Convert output files to input files (QUIP script `gap_prepare_sparsex_input.py gp.xml`) 
    * Optionally rename them to something shorter (e.g. `1.input`, `2.input` etc.)
4. Change input method (`sparse_method=FILE sparse_file=1.input`)
    * For more than one species use `add_species=F` and explicit input 
    * Check the number of braces (`gap={{$gap1}:{$gap2}:{$gap3}}`)
5. Run with `mpirun -np …`  (or `srun` for Slurm)
    * MPI can trip up on the `….xyz.idx` file. Usually works if it's there from the start, so retry can help.

**Comment by janklinux** [comment]
Thanks for the quick answer, I'll take some time any try it, reporting back when done.

**Comment by janklinux** [comment]
I have compiled using the recipe you provided and I have added the sparsify_only tag to the command line. So far gap_fit is running up to the point after the CUR decomposition, says 

Cartesian coordinates transformed to descriptors

libAtoms::Finalise: 2022-02-10 14:48:04
libAtoms::Finalise: Bye-Bye!

and exits normally. But I do not get any output files from this run besides the train.xyz.idx file, so there's no SparseX files I could use to go on further. Did I forget something?

**Comment by jameskermode** [comment]
Nothing in standard error? @Sideboard can you take a look at this?

**Comment by jameskermode** [comment]
Note that the option is `sparsify_only_no_fit` not `sparsify_only`. Can you provide the exact command line and input files you used?

**Comment by janklinux** [comment]
my fit script is:

```bash
#!/bin/bash

/home/kloppej1/compile/QUIP/build/linux_x86_64_gfortran_openmpi+openmp/gap_fit atoms_filename=train.xyz \\
   gap = { distance_2b Z1=78 Z2=78 compact_clusters=T cutoff=5.7 n_sparse=125 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform add_species=F : \\
           soap_turbo l_max=8 alpha_max={{8}} atom_sigma_r={{0.5}} atom_sigma_t={{0.5}} atom_sigma_r_scaling={{0.0}} atom_sigma_t_scaling={{0.0}} \\
           zeta=6 rcut_soft=5.2 rcut_hard=5.7 basis=poly3gauss scaling_mode=polynomial amplitude_scaling={{1.0}} n_species=1 species_Z={78} radial_enhancement={{1}} \\
           compress_file=compress.dat central_weight={{1.0}} add_species=F \\
           config_type_n_sparse={hcp:800:sc:800:bcc:800:fcc:2500:slab:2500:cluster:2500:dimer:80:addition:2500:elastic_small:100:elastic_medium:200:elastic_large:400:trimer:300:hkl:2500} \\
           delta=0.1 f0=0.0 covariance_type=dot_product sparse_method=cur_points } \\
 default_sigma={0.001 0.1 0.1 0.1} \\
 config_type_sigma={dimer:0.0001:0.01:0.01:0.01} \\
 sparsify_only_no_fit=T \\
 energy_parameter_name=free_energy force_parameter_name=forces force_mask_parameter_name=force_mask virial_parameter_name=virial sparse_jitter=1.0e-8 \\
 do_copy_at_file=F sparse_separate_file=T gp_file=platinum.xml openmp_chunk_size=10000
```

(edited to improve formatting) <- thanks ;) [The double \\ are actually single ones in my script, normal bash line continuation]

and besides that, there's no other output than what I already know from gap_fit, report on descriptors, partial derivs, target properties etc...

**Comment by Sideboard** [comment]
Could you please tell us which commit you have checked out? Just to be sure we are looking at the same version.

**Comment by janklinux** [comment]
The header is

libAtoms::Hello World: 2022-02-10 14:03:37
libAtoms::Hello World: git version  https://github.com/libatoms/QUIP.git,a7c7d98-dirty
libAtoms::Hello World: QUIP_ARCH    linux_x86_64_gfortran_openmpi+openmp
libAtoms::Hello World: compiled on  Feb 10 2022 at 13:57:47
libAtoms::Hello World: MPI parallelisation with 1 processes
libAtoms::Hello World: OpenMP parallelisation with 12 threads

git log, last entry:

commit a7c7d988ef055034566c1443b915e6c5e64a6b5a
Merge: bc02e46 ab8e64f
Author: James Kermode <james.kermode@gmail.com>
Date:   Wed Feb 9 13:18:55 2022 +0000

    Merge pull request #404 from Sideboard/fix/f2py_fcompiler
    
    fix: reconnect QUIPPY_FCOMPILER to f2py

**Comment by jameskermode** [comment]
The top of the git log for `QUIP/src/GAP` as well would be useful, just to be sure the submodule is pointing to the right place.

**Comment by janklinux** [comment]
cd src/GAP && git log, top entry:

commit b1f180464bda9048ab60442622cd9b90c4fb3feb
Author: Albert Bartok-Partay <apbartok@gmail.com>
Date:   Wed Sep 15 13:53:44 2021 +0100

    fixed bug in descriptor_calc_array(), which did not consider the masks correctly

**Comment by jameskermode** [comment]
Then you definitely need to `git pull` inside `src/GAP` as that's very old. I guess you had local modifications so the submodule didn't get updated correctly.

**Comment by janklinux** [comment]
I've had some time to try, reporting what I found now:
First, I have not been successful in running an MPI gap_fit job. While the sparse covariance matrix computation has a remarkable speed up compared to the openmp version, the code fails after that with 
On entry to PDGEQR2 parameter number    9 had an illegal value. 
I have tried different compilers (intel and gcc) and with openblas+netlib-scalapack and MKL and both give me the same error. Re-writing the descriptors to make sure they are consistent (with the sparsify_only_no_fit=T flag) did not help.

So at this point I'm stuck. I have experience with scalapack parallel code and I could take a look but I doubt I would make much progress without understanding it in more depth. Maybe you have some ideas what I could look at or try.

I am running this job on 2 nodes with 48 cores each. In the std.err there's no indication of a previous error, like memory problems etc (we have 3T physical memory per node), it's 
{32,0}, pnum=32, Contxt=0, killed other procs, exiting with error #1. and then
MPI_ABORT was invoked on rank 69 in communicator MPI_COMM_WORLD with errorcode 1.

**Comment by Sideboard** [comment]
Are these still the same versions as mentioned above? Still a7c7d98, which GAP version?

**Comment by janklinux** [comment]
QUIP:
commit a7c7d988ef055034566c1443b915e6c5e64a6b5a
Merge: bc02e46 ab8e64f
Author: James Kermode <james.kermode@gmail.com>
Date:   Wed Feb 9 13:18:55 2022 +0000

    Merge pull request #404 from Sideboard/fix/f2py_fcompiler
    
    fix: reconnect QUIPPY_FCOMPILER to f2py

GAP:
commit bf7b1ff6943f60765151fa9d0a656f9f9220542d
Merge: 91a7062 12c3d64
Author: albapa <apbartok@gmail.com>
Date:   Mon Feb 7 11:49:52 2022 +0000

    Merge pull request #27 from Sideboard/feat/ignore_forces
    
    feat: ignore forces if sparsify only

is what git log top entries show

**Comment by albapa** [comment]
I think we need a MWE or minimal failing example. Can we (@Sideboard) send a small example that works for us, and let @janklinux try it on their installation?

**Comment by Sideboard** [comment]
@janklinux Please update to the current QUIP and the coressponding GAP.

I assume you have quite a large calculation. In the earlier version the ScaLAPACK blocking sizes were set to the number of local rows. The work array for ScaLAPACK (length `lwork`) depends squared on the blocking size, so for large sizes you have to allocate more entries in the 1d work array than can be expressed by a 32bit integer.

In the new version, there is an argument `mpi_blocksize=<int>` that defaults to (value 0) the number of columns except when the number of rows is close to or smaller than that number (which happens when you have a large number of MPI processes). This is a compromise between keeping both the working array and the excess memory from the block size small.

Please try and tell if it works. Sorry for the inconvenience.

@albapa Concerning MWE, if quippy can be successfully built, `make test` includes three `gap_fit` tests  (`./tests/test_gapfit.py`), including a ScaLAPACK one. Maybe we could remove the dependency or add an additional, independent test.

**Comment by janklinux** [comment]
OK, I pulled with all submodules and recompiled. What should I use and value for `mpi_blocksize=<int>`? Is there a rule of thumb or just trial and error? I assume it's a gap_fit argument, similar to `sparsify_only_no_fit=T`?

**Comment by janklinux** [comment]
also ran into this now:

/home/kloppej1/compile/QUIP/src/GAP/gp_predict.f95:855:103:

          call SP_Matrix_QR_Solve(a, globalY, alpha, task_manager%n_workers, task_manager%ScaLAPACK_obj)
                                                                                                       1
Error: Type mismatch in argument 'scalapack_obj' at (1); passed INTEGER(4) to TYPE(scalapack)
make[1]: *** [gp_predict.o] Error 1
make[1]: Leaving directory `/scratch/work/kloppej1/compile/QUIP/build/linux_x86_64_gfortran_openmpi+openmp'
make: *** [GAP] Error 2

**Comment by Sideboard** [comment]
Error: Did you update both QUIP and GAP?

Value: You can set `mpi_blocksize` to any (positive) value you like. It will affect both running time and memory. The default is what I explained above, it's used if `mpi_blocksize=0` or if the argument is missing.

**Comment by janklinux** [comment]
I did git pull --recurse-submodules before recompilation, yes, that's when the error came up.

**Comment by Sideboard** [comment]
Then there are leftovers from the previous compilation, I guess. `make clean` to compile from scratch.

**Comment by janklinux** [comment]
I removed the build directory and started with make config
the error persists...

**Comment by Sideboard** [comment]
Ok, that is strange. Could you check the commit ids again, please? Then I can try to replicate.

**Comment by Sideboard** [comment]
Did `git pull --recurse-submodules` work or was it maybe blocked by local changes? `git submodule update` should work.

**Comment by janklinux** [comment]
The only way to get a consistent code version seems to be to rm -rf QUIP && git clone --recursive URL, otherwise I end up with partially dead code, or unmaintained versions of it. Neither git pull --recurse-submodules (which actually says it does stuff) nor git submodule update seems to get the proper job done. The last version did not even contain any local modifications, git just was not able to find the submodule of a submodule properly and I got a different compile time error, i.e. soap_turbo of GAP...

That said, I did a new recursive clone and it compiles now, gonna try the mpi argument and see how it goes.

**Comment by Sideboard** [comment]
Sorry for the trouble. These submodules are a pain sometimes. `git clone` is probably the easiest solution in such a case.

**Comment by janklinux** [comment]
Success!
I was able to train a potential on both our sumo nodes using 5T of physical memory ;)
The speedup is quite impressive, the previous potential filling a single 3T node to the brim was running 1day8hrs in 48-thread openmp mode, the MPI with almost double the memory usage took only 14hrs. 
Binary compiled with gcc/8.4 openmpi/4.0.5, linked to Intel MKL from conda.

Looking forward to new developments, good work!

**Comment by jameskermode** [comment]
Excellent news! We've seen good scaling up to ca. 2k cores so you might be able to push it even further!

**Comment by janklinux** [comment]
For us it's rather using our nodes to push for more memory which is easy enough now and was impossible in openmp. I am working on a sparsification method that writes out descriptors to use in gap_fit, similar to the ...sparsify_only... flag. This way we can specifically include some few configurations that are important while those tend to fall through the CUR.

**Comment by yuxzhou** [comment]
Hi all,

Many thanks to this fantastic test on openmp+openmpi version of gap_fit! I'm also trying to fit my GAP potential which requires more than 2TB memory and takes a rather long time (~1 week using 64 cores) to fit. I would very like to try this new version but just ran into issues telling me that 

```
Warning: Fortran 2018 deleted feature: DO termination statement which is not END DO or CONTINUE with label 175 at (1)
mpif90  -x f95-cpp-input -ffree-line-length-none -ffree-form -fno-second-underscore -fPIC -fno-realloc-lhs -fopenmp -I/public1/home/sc50138/Codes/QUIP-12.Mar/src/libAtoms -I/public1/home/sc50138/Codes/QUIP-12.Mar/src/fox/objs.linux_x86_64_gfortran_openmpi+openmp/finclude -O3  -DGETARG_F2003 -DGETENV_F2003 -DGFORTRAN -DFORTRAN_UNDERSCORE -D_MPI -D'GIT_VERSION="https://github.com/libAtoms/QUIP.git,2555f1f-dirty"' -D'GAP_VERSION=1645177290' -D'QUIP_ARCH="linux_x86_64_gfortran_openmpi+openmp"' -D'SIZEOF_FORTRAN_T=2' -DHAVE_GAP -DHAVE_PRECON -DHAVE_QR -DSCALAPACK  -c  /public1/home/sc50138/Codes/QUIP-12.Mar/src/libAtoms/error.f95 -o error.o
/public1/home/sc50138/Codes/QUIP-12.Mar/src/libAtoms/error.f95:45:6:

   45 |   use mpi
      |      1
Fatal Error: Cannot read module file ‘mpi.mod’ opened at (1), because it was created by a different version of GNU Fortran
compilation terminated.
make[1]: *** [error.o] Error 1
make[1]: Leaving directory `/public1/home/sc50138/Codes/QUIP-12.Mar/build/linux_x86_64_gfortran_openmpi+openmp'
make: *** [libAtoms] Error 2
```

I guess it's probably due to the wrong version of gfortran in my compilation. Do you have any suggestion of what to solve this?

Thanks so much for help!

**Comment by janklinux** [comment]
It looks like you are using the wrong mpif90. The mpi.mod is usually created when building the mpif90 so you should check that you are loading the right modules (on a cluster machine) and that you have the right include paths set. Check that mpif90 --version gives you the correct gfortran version when you call gfortran --version. That might be the issue for you.

**Comment by yuxzhou** [comment]
Thank you for the quick reply! I compiled the code with Scalapack/2.1.0 OpenBLAS/0.3.7 mpi/openmpi/4.1.1 and gcc/9.1.0. I also tried with gcc/8.3.0, but got the same error. I've checked that both `mpif90 --version` and `gfortran --version` provide exactly the same version of GNU fortran. Shall I look for the absolute path of mpi.mod and then include that in the source code?

**Comment by mcaroba** [comment]
@yuxzhou This doesn't answer your question, but if it takes 1 week on 64 cores that means you're 1) not using OpenMP acceleration or 2) using it inefficiently (set openmp_chunk_size=10000 or similar).

**Comment by janklinux** [comment]
Usually the correct version of mpi.mod (and other include headers) is automatically found by the wrapper script mpif90 so I don't know what happened to the MPI installation on your compute cluster. You can of course try to manually find and include the correct mpi.mod but this is only a workaround, not a fix, and you might run into other MPI related errors later. Maybe ask the admins of the compute cluster why this error appears so that they can fix the MPI installation.

**Comment by yuxzhou** [comment]
@mcaroba Thank you! Yes, the efficiency seems to be too slow that I proceed very slowly. As for OpenMP acceleration, I've chekced the running contidion in computing node using `top` and confirmed that the all CPU are in use (CPU% ~= 6400%). As for 2, shall I add the comment `openmp_chunk_size=10000` in my fitting command or any other place?

**Comment by janklinux** [comment]
You can add `openmp_chunk_size=10000` at the end of the gap_fit command line.

**Comment by noambernstein** [comment]
For what it’s worth, I’ve found than on our many core machines (56-72) it’s
actually faster to use only half the physical cores in OMP_NUM_THREADS

**Comment by yuxzhou** [comment]
@janklinux thanks for help! I've solved the mpi problem by loading the correct mpi and gcc modules! However, another issue just comes out:
```
/public1/home/sc50138/Codes/QUIP-12.Mar/src/libAtoms/task_manager.f95:354.13:

    allocate(this%idata, source=from%idata)
             1
Error: Array specification required in ALLOCATE statement at (1)
make[1]: *** [task_manager.o] Error 1
make[1]: Leaving directory `/public1/home/sc50138/Codes/QUIP-12.Mar/build/linux_x86_64_gfortran_openmpi+openmp'
make: *** [libAtoms] Error 2
```

Is there anything I've missed during the compilation?

**Comment by janklinux** [comment]
I had similar issues, it's most likely due to the submodules of git not all being up to date. Try getting a fresh QUIP clone and use git clone --recursive when doing so, same for turbogap. Then use all the same compilers and libraries, that should fix it.

**Comment by yuxzhou** [comment]
Thanks! Unfortunately, I `rm -rf QUIP` and tried `git clone --recursive https://github.com/libAtoms/QUIP.git`, but that doesn't help. The issue from `/src/libAtoms/task_manager.f95` would always come out. I also tried `git pull --recurse-submodules` and `git submodule update` after the git clone, however they doesn't help either. Is there any other suggestion for this problem or do I miss anything here? 

In fact, I've checked with my group members and they said that this issue came out every time when they compiled the latest version of QUIP even without using openmp+openmpi version. Do we have any method to confirm that the submodules are indeed up to date?

**Comment by noambernstein** [comment]
Looks like a Fortran compiler version problem to me. Maybe some default
standard has changed, and you need some compiler flag to turn it off

**Comment by noambernstein** [comment]
If it’s never worked on this system, could be too old a version (some web
searching on this error message suggests maybe it’s caused by old gfortran
versions).

**Comment by yuxzhou** [comment]
@noambernstein Thanks! Feel quite strange to me since I can compile QUIP using old source code but can't with the newly git cloned one. Maybe I could switch to some new fortran versions.

**Comment by janklinux** [comment]
A recursive clone is indeed that way to get the latest complete version. I did that a few days ago on a Laptop I bought just off the shelf, deleted all the windows remnants and installed Linux from scratch. I did a recursive pull and made all targets successfully. For me gcc/8.4 works on our compute cluster (also intel compiler works for openmp version, I haven't tried the intel-mpi yet) and gcc/11.2 works on my new laptop, too. Compilation and small test fits, all fine, openmp and MPI. 

P.S. Did you clone turbogap --recursive too? All versions have to be consistent...

**Comment by noambernstein** [comment]
Perhaps some new language feature is being used in the new version that
wasn't in the older one.

**Comment by janklinux** [comment]
Moika,

I want to report on some progress. I have done some testing on the MPI+OMP ratios on our SUMO machines (we have two 48CPUs+3T RAM) and other smaller nodes in MPI mode compared to the 'default' OMP version we have used so far. But before I report results that might be misleading I need to report something else:

I have two versions of gap_fit, a production version compiled with icc/ifort and linked to MKL with Intel OMP and a development version compiled with gcc/8.4 openmpi/4.0.5 and linked to that same Intel MKL but with OpenMPI+OMP. Let's call the first intel and the second gcc. Both stem from the same source tree, the difference is the QUIP_ARCH I export. I call the gap_fit binary with the full path from the build directory within QUIP, like /full/path/to/build/of/gap_fit in each job.

When I run the gcc version I get weird numbers in the report from gap_fit, namely the number of energies and forces is way off. I am using a rather small test set that has 9803 entries of energy and 160089 individual forces. This is reported correctly when I run the intel version, independent of the number of physical threads provided by SLURM and the number of OMP_NUM_THREADS I manually set. The Intel version is consistently reporting the correct number of energies and forces.

```
========== Report on number of target properties found in training XYZ: ========

Number of target energies (property name: free_energy) found: 9803
Number of target local_properties (property name: local_property) found: 0
Number of target forces (property name: forces) found: 160089
Number of target virials (property name: virial) found: 0
Number of target Hessian eigenvalues (property name: hessian) found: 0

================================= End of report ================================
```
It uses 460 GB RAM at this point.

The gcc version, however, does not. The report is consistent between different MPI+OMP configurations, i.e. 1MPI+48OMP, 12MPI+4OMP, 24MPI+2OMP and 48MPI+1OMP, but it states:

```
========== Report on number of target properties found in training XYZ: ========

Number of target energies (property name: free_energy) found: 117636
Number of target local_properties (property name: local_property) found: 0
Number of target forces (property name: forces) found: 1921068
Number of target virials (property name: virial) found: 0
Number of target Hessian eigenvalues (property name: hessian) found: 0

================================= End of report ================================
```

and the memory usage now is 1.2TB in total. This is clearly inconsistent to the OpenMP version. This is three times the memory in use compared to the Intel version and it is the same usage and report when I run the gcc version in single MPI mode, i.e. srun -n 1 <fit>. I am aware of MPI memory overhead but that's not the issue here. It's always that same memory amount independent of the number of MPI processes I choose. 

Also, I do write out my own SOAP descriptors to files and read them with gap_fit and I initially thought - well it's me again. But not this time... I was astonished to see the validation results from the gcc and Intel versions, they are the same, so the math reduces to the same result :O

Furthermore, I also tried to compile with Intel-MPI but this produces a binary that segfaults on execution. A long stack trace starting with some IntelMPI things (yes, I linked to mkl_blacs_intelmpi_lp64). All the reported things lead me to think there's some hidden issue in the MPI routines. It's probably well disguised and hiding in plain sight, as they always do ;)

Also, as if that is not enough already, I can fit a potential on one of our nodes with 48MPI+1OMP successfully. I do, however, get a segmentation fault after the covariance matrix computation, before it should say using ScaLapack to solve QR, when running the same input with 24MPI+2OMP and 12MPI+4OMP. And I don't understand why. The same input runs with 1MPI+48OMP gcc and Intel just fine and validates the same.

I am willing to do some more digging into this but we might want to talk in person to clarify efforts and directions of how to test and move forward, unless things seems clear to you and they are easily fixable.

**Comment by Sideboard** [comment]
@janklinux What QUIP and GAP versions were you using?

The property numbers are summed up via MPI and default to just the single input if there is only one process. So it should be consistent over all variations.

**Comment by janklinux** [comment]
I just did a fresh pull and now the reported numbers are consistent and memory usage looks more reasonable as well. The segfault issue on varying numbers of MPI+OMP combinations has not appeared with this pull either. I guess I caught some bad combination the last time. Maybe I'll find some time to test the Intel version too.

**Comment by Sideboard** [comment]
FYI: https://github.com/libAtoms/QUIP/pull/418 has been merged, i.e. the memory overhead is now reduced by having only one process compute Kmm (Lmm) and scattering the necessary rows via MPI.

**Comment by janklinux** [comment]
I can not fit a proper potential with the MPI version, I get NaN in the alphas or values in the range of 1E73, depending on the values I set for mpi_blocksize and openmp_chunk_size, haven't tried anything else both those two should be performance related only.
The gfortran does not compile but the ifort works and gives a proper potential when compiled in OMP.
git cloned --recursive on March 30th.

**Comment by albapa** [comment]
Can you share a minimal example, please?

**Comment by janklinux** [comment]
you mean the descriptors written to files, the train.xyz, the gap_fit string and the submit script?

**Comment by albapa** [comment]
Yes, but there is perhaps something you can do on your end: can you please run the tests? I'm wondering if that fails, too. Let us know.

**Comment by janklinux** [comment]
module purge && module load gcc/8.4 openmpi/4.0.5 intel-mkl/2019.3.199
export QUIP_ARCH=linux_x86_64_gfortran_openmpi+openmp
make test fails:
!! f90wrap patched version of f2py - James Kermode <james.kermode@gmail.com> !!

running build
running config_cc
INFO: unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
INFO: unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
INFO: build_src
INFO: building extension "_quippy" sources
INFO: f2py options: []
INFO: f2py:> ./src.linux-x86_64-3.9/_quippymodule.c
OSError: [Errno 2] No such file or directory: 'f90wrap_Atoms_types.f90'. Skipping file "f90wrap_Atoms_types.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_Atoms.f90'. Skipping file "f90wrap_Atoms.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_System.f90'. Skipping file "f90wrap_System.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_Dictionary.f90'. Skipping file "f90wrap_Dictionary.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_DynamicalSystem.f90'. Skipping file "f90wrap_DynamicalSystem.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_nye_tensor.f90'. Skipping file "f90wrap_nye_tensor.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_Potential.f90'. Skipping file "f90wrap_Potential.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_Potential_simple.f90'. Skipping file "f90wrap_Potential_simple.f90".
OSError: [Errno 2] No such file or directory: 'f90wrap_descriptors.f90'. Skipping file "f90wrap_descriptors.f90".
Reading f2cmap from '.f2py_f2cmap' ...
        Warning: redefinition of {'complex':{'8':'complex_float'->'complex_double'}}
        Mapping "complex(kind=8)" to "complex_double"
        Mapping "complex(kind=dp)" to "complex_double"
        Warning: redefinition of {'integer':{'':'int'->'int'}}
        Mapping "integer(kind=)" to "int"
        Warning: redefinition of {'integer':{'8':'long_long'->'long_long'}}
        Mapping "integer(kind=8)" to "long_long"
        Mapping "integer(kind=c_int)" to "long_long"
        Mapping "integer(kind=c_intptr_t)" to "long_long"
        Mapping "integer(kind=idp)" to "long_long"
        Warning: redefinition of {'real':{'16':'long_double'->'long_double'}}
        Mapping "real(kind=16)" to "long_double"
        Warning: redefinition of {'real':{'4':'float'->'float'}}
        Mapping "real(kind=4)" to "float"
        Warning: redefinition of {'real':{'8':'double'->'double'}}
        Mapping "real(kind=8)" to "double"
        Mapping "real(kind=DP)" to "double"
        Mapping "real(kind=c_double)" to "double"
        Mapping "real(kind=dp)" to "double"
        Mapping "real(kind=qp)" to "double"
Successfully applied user defined f2cmap changes
Reading fortran codes...
Post-processing...
Post-processing (stage 2)...
Building modules...
error: f2py target file './src.linux-x86_64-3.9/_quippymodule.c' not generated
make[1]: *** [quippy/_quippy.cpython-39-x86_64-linux-gnu.so] Error 1
make[1]: Leaving directory `/scratch/work/kloppej1/compile/MPI_QUIP/build/linux_x86_64_gfortran_openmpi+openmp'
make: *** [quippy] Error 2

make quippy fails with the same error.

**Comment by albapa** [comment]
There are clearly problems here, but not sure how they affect the rest of the build. Maybe @Sideboard can advise on how to run the gap_fit test by hand, that should be possible without quippy.

**Comment by Sideboard** [comment]
Something like
```
cd "$QUIP_ROOT/tests"
HAVE_GAP=1 HAVE_SCALAPACK=1 ./test_gapfit.py
```
should work. Can check again in the morning.

Edit: Forgot `BUILDDIR=$QUIP_ROOT/build/$QUIP_ARCH`.

**Comment by janklinux** [comment]
export BUILDDIR=/home/kloppej1/compile/MPI_QUIP/build/linux_x86_64_gfortran_openmpi+openmp
HAVE_GAP=1 HAVE_SCALAPACK=1 python test_gapfit.py
results in:
zsh: illegal hardware instruction  HAVE_GAP=1 HAVE_SCALAPACK=1 python test_gapfit.py

everything in compiled on the node I run it on, all the same modules.
module load gcc/8.4 openmpi/4.0.5 intel-mkl/2019.3.199 is the environment

**Comment by VladCarare** [comment]
Good morning; sorry to come in on the ongoing discussion - I'm not sure where else to type this. 

I am also attempting to run an mpi gap fit.

After successfully compiling it, it seems that whenever I run it with 'srun gap_fit ...' on Archer2, I get an error of a sorts: 
`srun: error: Allocation failure of 2 nodes: job size of 2, already allocated 2 nodes to previous components. `
where I asked for 2 nodes in the submission script. 
Should I have run the gap_fit with `srun --nodes=1 gap_fit ...`, then the error would be:
`srun: error: Allocation failure of 2 nodes: job size of 2, already allocated 1 nodes to previous components.`

I am supposing this error was introduced by some modifications made in the last commits, as my lab-mate is using mpi gap successfully. (not sure what commit he's using, but I am going to ask)

Top of my QUIP git log:

```
commit 53fa928b8bfb8d78d07bdcd7a0e0bcb7b8837798 (HEAD -> public, origin/public, origin/HEAD)
Merge: 5989901a5 fcb9531b9
Author: albapa <apbartok@gmail.com>
Date:   Wed Mar 30 14:08:36 2022 +0100

    Merge pull request #418 from Sideboard/fix/single_Lmm
    
    Fix: single Lmm
```


Top of my GAP git log:

```
commit 85ec22950ba4ca91c295c541cc8582d5124a3290 (HEAD)
Merge: 5ca681e 7681423
Author: albapa <apbartok@gmail.com>
Date:   Tue Mar 29 12:54:28 2022 +0100

    Merge pull request #33 from Sideboard/fix/single_Lmm
    
    Fix/single lmm
```

Edit: I mention that running it without srun works, but gives me only 1 mpi task.

**Comment by Sideboard** [comment]
@VladCarare I think these errors are not related to gap_fit but stem from some problem in the batch script. Could you please compare with your colleague to see if there are differences in your scripts?

**Comment by VladCarare** [comment]
I copied his script and am using the same submission, barring some of his own memory usage tracking bits. 

```
#!/bin/bash

#SBATCH --job-name=MPI-GAP-FIT
#SBATCH --nodes=2 --tasks-per-node=2 --cpus-per-task=64  # I tried many variations of these, and all fail with the same error
#SBATCH --time=04:20:00
#SBATCH --account=[...]
#SBATCH --partition=standard
#SBATCH --qos=standard

export OMP_NUM_THREADS=64
export PYTHONPATH="/work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/build/archer2-mpi+openmp/:$PYTHONPATH"
export PATH="/work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/build/archer2-mpi+openmp/:$PATH"


srun --distribution=block:block --hint=nomultithread gap_fit at_file=test.xyz \
    verbosity=VERBOSE mpi_print_all=T   sparse_jitter=1e-10     default_sigma={0.001 0.01 0.01 0.0}     energy_parameter_name=REF_energy     force_parameter_name=REF_forces    e0_method=isolated     do_copy_at_file=F     sparse_separate_file=T     gp_file=gap.xml     virial_parameter_name=REF_stress \
    gap={ Z={{16 16}} compact_clusters=T cutoff=5 cutoff_transition_width=1 distance_Nb=T order=2 covariance_type=ard_se f0=0.0 n_sparse=20 sparse_method=uniform theta_uniform=1 delta=1.7627404988182054 add_species=F : Z=16 atom_sigma=0.6 central_weight=1.0 cutoff=5 cutoff_transition_width=1.2 l_max=4 n_max=8 n_species=1 soap=T species_Z={{16}} covariance_type=dot_product f0=0.0 n_sparse=100 print_sparse_index=T sparse_method=cur_points zeta=4 delta=0.4281746205732923 add_species=F}
```

I'll try to find out whether he's on the same commit as I am (the last ones).


Also on a different but maybe related note, I'm in the tests folder and running test_gapfit.py returns me:
```
Traceback (most recent call last):
  File "test_gapfit.py", line 29, in <module>
    import quippytest
  File "/mnt/lustre/a2fs-work3/work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/tests/quippytest.py", line 22, in <module>
    import quippy
  File "/work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/build/archer2-mpi+openmp/quippy/__init__.py", line 2, in <module>
    import quippy._quippy
ImportError: /work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/build/archer2-mpi+openmp/quippy/_quippy.cpython-38-x86_64-linux-gnu.so: undefined symbol: GOMP_critical_end
```

**Comment by bernstei** [comment]
I agree that those are srun errors that (pretty much) cannot be related to what's inside gap_fit.  Can you run a different MPI program with that command line?  Do you have 128 physical cores on a node?

**Comment by bernstei** [comment]
I had some similar looking issues with mpirun when trying to use mixed OpenMP/MPI, but for me switching to srun fixed the problem. I wonder if there are still lurking issues even in srun.

**Comment by VladCarare** [comment]
> Can you run a different MPI program with that command line?  Do you have 128 physical cores on a node?

Yes and yes. For example the submission script below works - whereas the equivalent one for gap_fit does not work.

```
#!/bin/bash

#SBATCH --job-name=MPI-GAP-FIT
#SBATCH --nodes=2 --tasks-per-node=2 --cpus-per-task=64
#SBATCH --time=00:20:00

# Replace [budget code] below with your project code (e.g. t01)
#SBATCH --account=e89-came
#SBATCH --partition=standard
#SBATCH --qos=short

export OMP_NUM_THREADS=64

export PYTHONPATH="/work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/build/archer2-mpi+openmp/:$PYTHONPATH"
# export PYTHONPATH=${PYTHONPATH}:/opt/womble/QUIP/2021_02_08/linux_x86_64_gfortran_openmp/quippy
export PATH="/work/e89/e89/vc381/Calculations/dev/gap-mpi-2/QUIP/build/archer2-mpi+openmp/:$PATH"

module load castep
srun --nodes=2 --distribution=block:block --hint=nomultithread castep.mpi castep
```

**Comment by bernstei** [comment]
@Sideboard does the MPI gap_fit do anything nontrivial with the MPI processes, e.g. `MPI_Comm_Spawn` or something like that? Or is it just using a fixed MPI universe, starting from `MPI_COMM_WORLD` and maybe doing comm splits?

**Comment by Sideboard** [comment]
@bernstei Only MPI_COMM_WORLD is used.

**Comment by VladCarare** [comment]
I spoke with my colleague - for whom the same commands seem to work. He is using previous commits for QUIP `commit 5989901a55f8b5c847c555daa9759115f52e2c12`; and for GAP `commit 3b712bf001f23cca3a48eb1293f356c0766f6b41`.

**Comment by bernstei** [comment]
Any chance the changed qos is leading it to select different nodes?  I'm fairly confident that this error happens _before_ the process actually starts, so I don't see how it can be dependent on the executable.

Also, one thing which is weird is that the error message comes from the slurm code related to "heterogeneous" jobs.  I'm not sure why that would even be triggered here.

Aha - I have an idea. Recent versions of slurm can request different sets of resources within a single allocation somehow. Those are separated by `:`.  Is there any chance the `:` in the gap descriptor command line are confusing `srun`?  Can you hide those from srun, either by using the reading from a file option, or by wrapping your particular `gap_fit` command line in a script, e.g. `temp_script.sh`, and `srun --nodes=2 --distributions... ./temp_script.sh` ?

**Comment by VladCarare** [comment]
> Aha - I have an idea. Recent versions of slurm can request different sets of resources within a single allocation somehow. Those are separated by :. Is there any chance the : in the gap descriptor command line are confusing srun? Can you hide those from srun, either by using the reading from a file option, or by wrapping your particular gap_fit command line in a script, e.g. temp_script.sh, and srun --nodes=2 --distributions... ./temp_script.sh ?

This is indeed the reason my script wasn't working @bernstei . Thanks!


In fact this was already in the steps outlined by James Kermode in the first few comments: 

> Check the number of braces (gap={{$gap1}:{$gap2}:{$gap3}})

I had not put the curly braces for my 2b and my soap descriptors. Doing it also solves the error.

**Comment by Sideboard** [comment]
Concerning the original topic of this issue:

I've had a chat with @janklinux and we found that the current MPI version of gap_fit needs an explicit number of sparse points for each descriptor. The automatic recognition from the file length is done after the distribution. Until this is fixed, you should use an explicit number.

**Comment by janklinux** [comment]
I have tested what we talked about @Sideboard but it did not fix the problem. The missing sparse points might still be the underlying issue, but specifying them in the fit command did not change the behavior. I still get NaN in the alphas when running it with more than one MPI process while OMP versions work with and without explicit sparse point declaration. Also the MPI gives the correct potential when run with -n 1.
On our cluster I also got 
`WARNING: get_shared_task_counts: Not enough data. Were sparse points reduced since task distribution?`
when I set the config_type_n_sparse manually. Remember that I can not specify them because I don't know from which config type they came from due to me selecting them myself. So I set an arbitrary value about which to code did not complain anyway, it then reports covariances not to be zero anymore in the memory summary but the resulting NaNs don't change. Also, the OMP version does report covariances to be zero as well while giving a proper potential, just for completeness.

**Comment by tavanfran** [comment]
Dear @jameskermode 
regarding the MPI: I have some GAP files ready for input in Lammps from the already published data (eg. for HfO2 and GeSbTe). I was wondering if it is possible to use the gap_prepare_sparsex_input.py (or another script) to reduce the size of these input files.

Another question: in your first comment on this topic, from what I understand, I have to first run a gap_fit with sparsify_only_no_fit=T, then another gap_fit with sparsify_only_no_fit=F and all the explicit input for each species. Is it correct? I tried this approach and the memory usage is much lower respect to the direct fit. Is this correct?

Thank you

**Comment by Sideboard** [comment]
@tavanfran `gap_prepare_sparsex_input.py` reads in the `gp.xml` output from a `gap_fit` run and the corresponding `gp.xml.*` files and just adds the cutoff radius for each sparseX to create the `gp.xml.*.input` files. These are needed to use the `sparse_method=file sparse_file=name.input` needed for MPI (but also usable for non-MPI runs).

`sparsify_only_no_fit=T` is recommended to create the `gp.xml.*.input` files except if you already have the sparseX points. This run does not allocate derivatives and stops right after choosing the sparse points, so it is fast and has a very small memory footprint. You will see that the xprime entries in the memory estimate will have 0 entries. The estimate is not adapted for the `sparsify_only_no_fit=T`, however, so don't read too much out of it in that case. Also note that the memory estimate in the full calculation only shows the estimate for the first process on the first node.

**Comment by janklinux** [comment]
I again have trouble running the MPI gap fit, I did a fresh recursive clone this morning. Problem is as follows:
Fitting a rather small potential (small database of about 80MB train.xyz) works and validates to the same result as running the single core compiled version in openmp or running the MPI gap_fit on a single core with threads. This is expected behavior I think. 
Now, running a larger database with 360MB train.xyz the MPI version does not complete the fit. Both the openmp and the MPI version on a single core with N threads work and again validate to the same result but the MPI version with more than 1 MPI core dies after the message

Using ScaLAPACK to solve QR
Program received signal SIGSEGV: Segmentation fault - invalid memory reference. (and a lengthy backtrace)

Peak memory usage is 900GB from 3TB, so just about 1/3 of the node, that can't be the problem here, can it?
If you have working MPI versions, I can provide my input if needed. From my testing it is related to the problem size, i.e. MBs of train.xyz. I don't have an exact file size tho. 
For all my fitting I select and write out the descriptors myself and that works just fine for the single core + openmp and the smaller MPI runs.
Any ideas what could be going wrong?

P.S. I have noticed a drastic reduction in memory use when running gap_fit, great work, now I can add even more data ;)

**Comment by Sideboard** [comment]
My first idea is that some array size is beyond the 32bit integer limit. This should be caught before the covariances are calculated but maybe we didn't consider every case.

- What are the sizes of your fit run, i.e. number of descriptors, target properties etc.?
- What does the memory estimate say? (It also includes the sizes of the matrices.)

**Comment by janklinux** [comment]
========================= Memory Estimate (per process) ========================

Descriptors
Descriptor 1 :: x 1 13268130 memory 106 MB
Descriptor 1 :: xPrime 1 236499396 memory 1891 MB
Descriptor 2 :: x 72 280739 memory 161 MB
Descriptor 2 :: xPrime 72 120739077 memory 69 GB
Subtotal 71 GB

Covariances
yY 125 863831 memory 863 MB * 2
yy 125 125 memory 125 KB
A 125 863956 memory 863 MB * 2
Subtotal 3455 MB

Peak1 72 GB
Peak2 3455 MB
PEAK  72 GB

Free system memory  3160 GB
Total system memory 3245 GB

**Comment by janklinux** [comment]
but also

===================== Report on number of descriptors found ====================

Descriptor 1:  distance_2b Z1=78 Z2=78 compact_clusters=T cutoff=5.7 n_sparse=125 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=file sparse_file=1.in add_species=F
Number of descriptors:                        106328414
Number of partial derivatives of descriptors: 1895124996

Descriptor 2:  soap_turbo l_max=8 alpha_max={8} atom_sigma_r={0.5} atom_sigma_t={0.5} atom_sigma_r_scaling={0.0} atom_sigma_t_scaling={0.0} zeta=6 rcut_soft=5.2 rcut_hard=5.7 basis=poly3gauss scaling_mode=polynomial amplitude_scaling={1.0} n_species=1 species_Z=78 radial_enhancement={1} compress_mode=trivial central_weight={1.0} add_species=F delta=0.1 f0=0.0 covariance_type=dot_product sparse_method=file sparse_file=Pt.soap
Number of descriptors:                        2245926
Number of partial derivatives of descriptors: 967476744

**Comment by tcnicholas** [comment]
Was this issue resolved?

For particularly large database/nsparse configurations, my fits compute the covariance matrices successfully, but then I get a segmentation fault at the `Using ScaLAPACK to solve QR` point of the program, followed by a lengthy backtrace.

When I used ~100 less configurations in my database, and fewer sparse points, this fit completed successfully.

This is with the following memory estimates:

```Descriptors
Descriptor 1 :: x 1 625718 memory 5005 KB
Descriptor 1 :: xPrime 1 11262924 memory 90 MB
Descriptor 2 :: x 1 625718 memory 5005 KB
Descriptor 2 :: xPrime 1 11262924 memory 90 MB
Descriptor 3 :: x 1 625718 memory 5005 KB
Descriptor 3 :: xPrime 1 11262924 memory 90 MB
Descriptor 4 :: x 1 625718 memory 5005 KB
Descriptor 4 :: xPrime 1 11262924 memory 90 MB
Descriptor 5 :: x 1 625718 memory 5005 KB
Descriptor 5 :: xPrime 1 11262924 memory 90 MB
Descriptor 6 :: x 1 625718 memory 5005 KB
Descriptor 6 :: xPrime 1 11262924 memory 90 MB
Descriptor 7 :: x 1 625718 memory 5005 KB
Descriptor 7 :: xPrime 1 11262924 memory 90 MB
Descriptor 8 :: x 1 625718 memory 5005 KB
Descriptor 8 :: xPrime 1 11262924 memory 90 MB
Descriptor 9 :: x 3 11958 memory 286 KB
Descriptor 9 :: xPrime 3 322866 memory 7748 KB
Descriptor 10 :: x 3 11958 memory 286 KB
Descriptor 10 :: xPrime 3 322866 memory 7748 KB
Descriptor 11 :: x 3 11958 memory 286 KB
Descriptor 11 :: xPrime 3 322866 memory 7748 KB
Descriptor 12 :: x 3 11958 memory 286 KB
Descriptor 12 :: xPrime 3 322866 memory 7748 KB
Descriptor 13 :: x 3 57403 memory 1377 KB
Descriptor 13 :: xPrime 3 1549881 memory 37 MB
Descriptor 14 :: x 3 57403 memory 1377 KB
Descriptor 14 :: xPrime 3 1549881 memory 37 MB
Descriptor 15 :: x 1500 2868 memory 34 MB
Descriptor 15 :: xPrime 1500 1997685 memory 23 GB
Descriptor 16 :: x 2640 2868 memory 60 MB
Descriptor 16 :: xPrime 2640 2042244 memory 43 GB
Descriptor 17 :: x 2640 1912 memory 40 MB
Descriptor 17 :: xPrime 2640 1347471 memory 28 GB
Descriptor 18 :: x 1500 478 memory 5736 KB
Descriptor 18 :: xPrime 1500 317196 memory 3806 MB
Subtotal 100 GB

Covariances
yY 1300 24525 memory 255 MB * 2
yy 1300 1300 memory 13 MB
A 1300 25825 memory 268 MB * 2
Subtotal 1060 MB

Peak1 100 GB
Peak2 1060 MB
PEAK  100 GB

Free system memory  522 GB
Total system memory 540 GB
```

---
QUIP version:

libAtoms::Hello World: git version  https://github.com/libAtoms/QUIP.git,v0.9.12-11-gbebe7183b-dirty
libAtoms::Hello World: QUIP_ARCH    archer2_mpich+openmp
libAtoms::Hello World: compiled on  Feb 13 2023 at 16:10:57

**Comment by janklinux** [comment]
yes, it was resolved and works now, I think it got sidetracked into the long integer issue and solved there. Anyway, it's fixed, thanks for the good work.

**Comment by tcnicholas** [comment]
(sorry to reopen this). Should this be working in the QUIP version I reported above? Because I am experiencing the same problem stilll?

**Comment by albapa** [comment]
Hi @tcnicholas , does the backtrace give any useful pointers on where to look at?

**Comment by tcnicholas** [comment]
This is the output:

```
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Backtrace for this error:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2b91a7f4549f in ???
#1  0x2b91a3dba1f3 in ???
srun: error: nid006501: task 0: Segmentation fault (core dumped)
srun: launch/slurm: _step_signal: Terminating StepId=3666124.0
slurmstepd: error: *** STEP 3666124.0 ON nid006501 CANCELLED AT 2023-05-17T10:01:03 ***
```

**Comment by janklinux** [comment]
the current version I have does not SEGFAULT anymore but instead stops and reports that the number of CPUs chosen for the job is too small and also reports the minimum number of CPUs needed to run properly. Maybe do a complete recursive QUIP clone and try again?

**Comment by tcnicholas** [comment]
Okay, I will run the same tests with the latest version of QUIP and report back. Thanks!

**Comment by rxk900** [comment]
Following the above steps (mirrored in https://libatoms.github.io/GAP/gap_fit.html#running-with-mpi)

I get  at Step 4

>  count_entries_in_sparse_file:  does not exist

If it helps header is:
libAtoms::Hello World: 2024-07-18 17:40:40
libAtoms::Hello World: git version  https://github.com/libAtoms/QUIP.git,v0.9.12-14-gb55e89d55-dirty
libAtoms::Hello World: QUIP_ARCH    linux_x86_64_ifort_icc_mpi
libAtoms::Hello World: compiled on  Jul 18 2024 at 11:08:53
libAtoms::Hello World: MPI parallelisation with 4 processes

I can see it should come from GAP/gp_fit.f95 and I can see ../build/linux_x86_64_ifort_icc_mpi/gp_fit.o but have run out of genki and thought I may as well ask if it has been seen before.

I have come back to this after having given up and gone to gfortran and OpenMP for an easier life which worked for gap_fit but that broke my LAMMPS (which is Intel OpenMPI based) so I am going back to trying the various combinations until I get something to work. 

My hunch is that it is a gfortran thing (ultimately - above message looks like a makefile thing) because the container downloads work for a colleague but working at an HPC center we are encouraged to make our own architecture customised builds.

**Comment by gabor1** [comment]
Why would it be a gfortran thing when you are using ifort? In fact is that the problem? If your mpi is set up with gfortran you should use gfortran with gap too

**Comment by rxk900** [comment]
No, I said ultimately (the LAMMPS fail). I returned to the intel MPI build when I decided to start again because I realised my first fail was an IntelMPI vs OpenMPI mismatch.

So far I have 
Intel build with OpenMPI  (first failed because needs SCALAPACK) and then failed with an error which I now realise was I had IntelMPI lines somewhere
Then I went to Intel OpenMP which ran gap_fit fine but LAMMPS is failing with 
[gadi-login-03:312241:0:312241] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x7ffcf48c3d88)
==== backtrace (tid: 312241) ====
 0 0x0000000000012cf0 __funlockfile()  :0
 1 0x0000000000bde666 potential_simple_module_mp_potential_simple_calc_()  /scratch/x89/rxk900/ML-IP/QUIP/src/Potentials/Potential_simple.f95:375
 2 0x0000000000a86ad9 potential_module_mp_potential_calc_()  /scratch/x89/rxk900/ML-IP/QUIP/src/Potentials/Potential.f95:937
 3 0x0000000000a98332 potential_module_mp_potential_sum_calc_()  /scratch/x89/rxk900/ML-IP/QUIP/src/Potentials/Potential_Sum_routines.f95:95
 4 0x0000000000a86954 potential_module_mp_potential_calc_()  /scratch/x89/rxk900/ML-IP/QUIP/src/Potentials/Potential.f95:940
 5 0x00000000009b4361 quip_lammps_wrapper()  /scratch/x89/rxk900/ML-IP/QUIP/src/Potentials/quip_lammps_wrapper.f95:147
 6 0x000000000082d31a LAMMPS_NS::PairQUIP::compute()  /scratch/x89/rxk900/3Aug2022/lammps-3Aug2022-QUIP/src/Obj_openmpi/../pair_quip.cpp:151
 7 0x00000000005d918e LAMMPS_NS::Verlet::setup()  /scratch/x89/rxk900/3Aug2022/lammps-3Aug2022/src/Obj_openmpi/../verlet.cpp:140
 8 0x00000000007e9f40 LAMMPS_NS::Run::command()  /scratch/x89/rxk900/3Aug2022/lammps-3Aug2022/src/Obj_openmpi/../run.cpp:171
 9 0x0000000000416fb7 LAMMPS_NS::Input::execute_command()  /scratch/x89/rxk900/3Aug2022/lammps-3Aug2022/src/Obj_openmpi/../input.cpp:845
10 0x0000000000414218 LAMMPS_NS::Input::file()  /scratch/x89/rxk900/3Aug2022/lammps-3Aug2022/src/Obj_openmpi/../input.cpp:301
11 0x00000000004118ec main()  /scratch/x89/rxk900/3Aug2022/lammps-3Aug2022/src/Obj_openmpi/../main.cpp:98
12 0x000000000003ad85 __libc_start_main()  ???:0
13 0x00000000004117ae _start()  ???:0

Following the LAMMPS fail decided to then try gfortran with OpenMPI and got distracted when I realised I had linked in IntelMPI at some point so went back to initial (build of choice) Intel with OpenMPI and got the count_entries error.

**Comment by rxk900** [comment]
I was wondering if it is a gfortran thing because my friend's singularity pull worked straight out of the box (or I guess container). Our MPI is set up for Intel and gfortran - we have a wrapper that picks up the appropriate one for the loaded compiler.
I am about to test out my hypothesis by going back to the gfortran build but ideally would prefer to get the intel compiler build working.

**Comment by albapa** [comment]
The code you use for training and evaluation does not have to be compiled with the same compiler. I.e. you can use an intel-compiled `gap_fit` and a gfortran-compiled LAMMPS to run your potential and it will be fine, as the model file is an XML.

Another thing that you need to clarify is that
- when training, you can use serial, openmp, MPI-compiled, and openmp/MPI code. All will work, but obviously the parallelisation will be different. The compiler should not matter as long as it's consistent with your MPI libraries.
- when using from LAMMPS, you can only use QUIP that you compiled in serial.

Regarding the instructions at https://libatoms.github.io/GAP/gap_fit.html#running-with-mpi you should verify if the input files generated in step 2 exists in the directory you're working in, and that you call them consistently in the command line (step 3).

Please note that a one-step training with MPI should also work thanks to @Sideboard

**Comment by rxk900** [comment]
10pm here so will get back to this tomorrow. All my LAMMPS (and the ones we provide on our supercomputer) are intel compiler and openmpi  based because our guys put most effort into performance optimising those so I wanted to stick with that. I haven't got round to gfortran yet - that got distracted.
The one-step thing still gives me:
================================= End of report ================================

SYSTEM ABORT: proc=1 Only sparse_method FILE implemented for MPI.   
SYSTEM ABORT: proc=2 Only sparse_method FILE implemented for MPI.   
SYSTEM ABORT: proc=3 Only sparse_method FILE implemented for MPI.   
SYSTEM ABORT: proc=0 Only sparse_method FILE implemented for MPI.   
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 1 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------

Oh, and I think file files were generated if we're talking 1.input etc (my notes are at work) and the non-existence count_entries_in_sparse_file didn't seem to be a file but a function. Also, from memory and it's getting late. This was meant to be a quick message to say the 1-steo thing didn't work.

**Comment by albapa** [comment]
> All my LAMMPS (and the ones we provide on our supercomputer) are intel compiler and openmpi based
You can (and should) compile LAMMPS with MPI. But the QUIP library you link to **must** be compiled in serial. This is documented in the readme.

> Oh, and I think file files were generated if we're talking 1.input etc (my notes are at work) and the non-existence count_entries_in_sparse_file didn't seem to be a file but a function.

The error `count_entries_in_sparse_file: does not exist` means that the sparse file you specified does not exist. In fact, it does report back the file name, but in your case it seems to be empty (i.e. it should be something like

`count_entries_in_sparse_file: blahblah does not exist`

What this tells me that in the command line the keyword `sparse_file` is not specified for every descriptor.

**Comment by bernstei** [comment]
> but in your case it seems to be empty

we could consider adding quotation marks around the filename in the error printing message, so it's obvious there's a blank string there:
`count_entries in sparse_file: '' does not exist`

**Comment by rxk900** [comment]
Ah, that is helpful. yes the quote marks would have helped. So it is missing files - or my input - in which case I have

Step 1 makes
-rw-r--r--  1 rxk900 x89 15173196 Jul 18 17:15 GAP.xml
-rw-r--r--  1 rxk900 x89      405 Jul 18 17:15 GAP.xml.sparseX.GAP_2024_7_18_600_17_15_41_3091
-rw-r--r--  1 rxk900 x89     4050 Jul 18 17:15 GAP.xml.sparseX.GAP_2024_7_18_600_17_15_41_3092
-rw-r--r--  1 rxk900 x89   809216 Jul 18 17:15 GAP.xml.sparseX.GAP_2024_7_18_600_17_15_41_3093

Step 2. Convert output files to input files ($QUIP_ROOT/bin/gap_prepare_sparsex_input.py gp.xml).
  Optional: Rename them to something shorter (e.g. 1.input, 2.input etc.).
../bin/gap_prepare_sparsex_input.py GAP.xml 
makes
-rw-rw----  1 rxk900 z00      695 Jul 18 17:26 GAP.xml.sparseX.GAP_2024_7_18_600_17_15_41_3091.input
-rw-rw----  1 rxk900 z00     5005 Jul 18 17:26 GAP.xml.sparseX.GAP_2024_7_18_600_17_15_41_3092.input
-rw-rw----  1 rxk900 z00   813016 Jul 18 17:26 GAP.xml.sparseX.GAP_2024_7_18_600_17_15_41_3093.input
and renamed to 1.input, 2.input etc.
3. Change input method (sparse_method=FILE sparse_file=1.input).
  For more than one species use add_species=F and explicit input.
  Check the number of braces (gap={{$gap1}:{$gap2}:{$gap3}}).

I wasn't sure about this bit. What is the syntax for more than one descriptor? I only have one species and my input isn't of the form gap={{$gap1}:{$gap2}:{$gap3}}). Oh I see. Now I think I understand what  {$gap1}:{$gap2}:{$gap3}} means. That's not for  multiple species - that is gap1 <- sparse_method=FILE sparse_file=1.input : gap2 <-  sparse_method=FILE sparse_file=2.input : gap3 <-  sparse_method=FILE sparse_file=3.input.

It's now past midnight so I think I'll try tomorrow but can you give an expanded example of what gap={{$gap1}:{$gap2}:{$gap3}} looks like?

**Comment by albapa** [comment]
> > but in your case it seems to be empty
> 
> we could consider adding quotation marks around the filename in the error printing message, so it's obvious there's a blank string there: `count_entries in sparse_file: '' does not exist`

Good suggestion. Now done in https://github.com/libAtoms/QUIP/pull/659

**Comment by albapa** [comment]
>but can you give an expanded example of what gap={{$gap1}:{$gap2}:{$gap3}} looks like?

It's exactly like you have written above. If you have only one specie, it's easy: you add the appropriate sparse file to every descriptor. If you have multiple species, it becomes tedious: you have to manually create all descriptors (in practice copy the ones created automatically by QUIP in the previous step), and decorate each of them with `sparse_method=FILE sparse_file=n.input`. Best done in a script.

Eliminating this was the purpose of https://github.com/libAtoms/QUIP/pull/581 which I am surprised doesn't work for you. Maybe if you can send a minimal failing example (command line, input files etc.) we can investigate.

**Comment by rxk900** [comment]
Not quite sure I have understood because my interpretation of gap={{$gap1}:{$gap2}:{$gap3}}) gives
RMAPS found multiple applications to be launched, with
at least one that failed to specify the number of processes to execute.
When specifying multiple applications, you must specify how many processes
of each to launch via the -np argument.

This input isn't mine to pass on - just trying to get stuff to work in my role as computer support - I could make a toy example but maybe that can wait till after weekend. Sorry, didn't post this but output is on desktop so I will have to switch to that.

**Comment by rxk900** [comment]
So toy example:
$ gap_fit energy_parameter_name=energy force_parameter_name=forces do_copy_at_file=F sparse_separate_file=T gp_file=GAP.xml at_file=train.xyz default_sigma=\{0.008 0.04 0 0\} gap=\{distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform add_species=T n_sparse=10\}
works
$ mpirun -np 4 gap_fit_mpi energy_parameter_name=energy force_parameter_name=forces do_copy_at_file=F sparse_separate_file=T gp_file=GAP.xml at_file=train.xyz default_sigma=\{0.008 0.04 0 0\} gap=\{distance_2b cutoff=4.0 covariance_type=ard_se delta=0.5 theta_uniform=1.0 sparse_method=uniform add_species=T n_sparse=10\}
gives 
SYSTEM ABORT: proc=0 Only sparse_method FILE implemented for MPI.   
SYSTEM ABORT: proc=2 Only sparse_method FILE implemented for MPI.   
SYSTEM ABORT: proc=3 Only sparse_method FILE implemented for MPI.   
SYSTEM ABORT: proc=1 Only sparse_method FILE implemented for MPI.

I'll now go back to trying the 4-step workround for this example.  It would help if that page (https://libatoms.github.io/GAP/gap_fit.html) had a working example - input and command line.

**Comment by rxk900** [comment]
And the 4-step workround works for this example.  I'll go back to my real example - at least I know I have the input syntax right.
I also think I found a typo along the way which may explain the previous error though not sure how it would have manifested as am RMAPS error.

**Comment by rxk900** [comment]
Haven't got the RMAPS error again but doing exactly the same steps as for Tutorial 3 for my actual case is failing with the error:
SYSTEM ABORT: proc=3 Traceback (most recent call last)
File "System.f95", line 1471 kind unspecified
string_to_real: could not convert, iostat=59  

which according to the non-MPI output is probably at the parsing descriptor stage (last thing to print is Found 3 GAPs):
Found 3 GAPs.
Descriptors have been parsed

I think I'll move on and go back to the OpenMP version. I feel I gave it my best shot.

**Comment by albapa** [comment]
It would still help if you pasted here the command line that fails.

**Comment by rxk900** [comment]
Not an input to share publicly I feel. Not even sure I feel comfortable sharing privately....
But I guess command line is harmless
mpirun -np 4 gap_fit_mpi at_file=train.xyz \
gap=\{\{distance_Nb order=2 \
                 cutoff=5.0 \
                 covariance_type=ARD_SE \
                 theta_uniform=1.0 \
                 n_sparse=15 \
                 delta=1.0:\
     distance_Nb 13 order=3 \
                    cutoff=4.0 \
                    covariance_type=ARD_SE \
                    theta_uniform=1.0 \
                    n_sparse=50 \
                    delta=0.004:\
     soap cutoff=4.0 \
          covariance_type=dot_product \
          zeta=2 \
          delta=0.016 \
          atom_sigma=0.7 \
          l_max=6 \
          n_max=6 \
          n_sparse=200 \
          sparse_method=FILE sparse_file=1.input\} \
:\{distance_Nb order=2 \
                 cutoff=5.0 \
                 covariance_type=ARD_SE \
                 theta_uniform=1.0 \
                 n_sparse=15 \
                 delta=1.0:\
     distance_Nb 13 order=3 \
                    cutoff=4.0 \
                    covariance_type=ARD_SE \
                    theta_uniform=1.0 \
                    n_sparse=50 \
                    delta=0.004:\
     soap cutoff=4.0 \
          covariance_type=dot_product \
          zeta=2 \
          delta=0.016 \
          atom_sigma=0.7 \
          l_max=6 \
          n_max=6 \
          n_sparse=200 \
          sparse_method=FILE sparse_file=2.input\} \
:\{distance_Nb order=2 \
                 cutoff=5.0 \
                 covariance_type=ARD_SE \
                 theta_uniform=1.0 \
                 n_sparse=15 \
                 delta=1.0:\
     distance_Nb 13 order=3 \
                    cutoff=4.0 \
                    covariance_type=ARD_SE \
                    theta_uniform=1.0 \
                    n_sparse=50 \
                    delta=0.004:\
     soap cutoff=4.0 \
          covariance_type=dot_product \
          zeta=2 \
          delta=0.016 \
          atom_sigma=0.7 \
          l_max=6 \
          n_max=6 \
          n_sparse=200 \
          sparse_method=FILE sparse_file=3.input\}\} \
e0_method=average \
default_sigma=\{0.001 0.5 0.0 0.0\} \
gp_file=GAP.xml

**Comment by gabor1** [comment]
why do you have "13" after the token "distance_Nb" ?

**Comment by rxk900** [comment]
Ah, that is a rogue line-number (I was given the input as a code block embedded in pdf which I C&P'd and stripped line numbers). Bizarre thing was that it seemed to work in the non-MPI version after my first MPI attempt failed with the IntelMPI/OpenMPI mismatch so I thought input and command was ok. I'll delete that and try again but that will have to wait till tomorrow. Sorry about the careless mistake.

**Comment by rxk900** [comment]
With the 13 removed I get the same error (but I have high hopes it will explain the original LAMMPS failure). I will move on now going back to my OpenMP build. (This was just a performance comparison diversion for me).

**Comment by albapa** [comment]
I've just noticed that you have extra {} around the individual descriptors. There shouldn't be - they are separated by the `:`.

**Comment by albapa** [comment]
I'll send you another line shortly, because that's not the only problem with it.

**Comment by albapa** [comment]
I haven't tried this because I don't have your training file, but this should be a good place to start.
```
mpirun -np 4 gap_fit_mpi at_file=train.xyz
gap={distance_Nb order=2
cutoff=5.0
covariance_type=ARD_SE
theta_uniform=1.0
n_sparse=15
delta=1.0 sparse_method=FILE sparse_file=1.input:
distance_Nb order=3
cutoff=4.0
covariance_type=ARD_SE
theta_uniform=1.0
n_sparse=50
delta=0.004 sparse_method=FILE sparse_file=2.input:
soap cutoff=4.0
covariance_type=dot_product
zeta=2
delta=0.016
atom_sigma=0.7
l_max=6
n_max=6
n_sparse=200
sparse_method=FILE sparse_file=3.input}
e0_method=average
default_sigma={0.001 0.5 0.0 0.0}
gp_file=GAP.xml
```
